### PR TITLE
Core/PathFinding: Add support to WorldObject pathfinding

### PR DESCRIPTION
**Pull request by jackpoz**
**Changes proposed:**

-  Core/PathFinding: Add support to WorldObject pathfinding

**Target branch(es):** 3.3.5/master

- [x] 3.3.5
- [ ] master

**Issues addressed:** None

**Tests performed:**
@Ovahlord tested it

**Known issues and TODO list:** (add/remove lines as needed)

None

### CODE REVIEW

**Comment by illfated** [comment]
Interesting to see this PR in combination with your other PR, the raycast one. Where (or when) will the improvements be most noticeable while playing?

**Comment by jackpoz** [comment]
I did these 2 PRs for @Ovahlord some weeks ago (see the commit dates), he was fiddling with some spell effects. Ask him for the full explaination and maybe some screenshots too

**Comment by Ovahlord** [comment]
The results will be easy when done: Spells like blink, random destination targets etc use a method to detect walls and obstacles and change their destination based on it.

The short version: that method straightup sucks in its current inplementation. We are going to change it to use mmap data to get proper collision data which will lead to safer and improved obstacle detections.

**Comment by illfated** [comment]
Thank you.

File: src/server/game/Movement/PathGenerator.cpp [hunk 1]
```diff
@@ -72,11 +72,12 @@ bool PathGenerator::CalculatePath(float destX, float destY, float destZ, bool fo
     _forceDestination = forceDest;
     _straightLine = straightLine;
 
-    TC_LOG_DEBUG("maps.mmaps", "++ PathGenerator::CalculatePath() for %u", _sourceUnit->GetGUID().GetCounter());
+    TC_LOG_DEBUG("maps.mmaps", "++ PathGenerator::CalculatePath() for %u", _source->GetGUID().GetCounter());
 
     // make sure navMesh works - we can run on map w/o mmap
     // check if the start and end point have a .mmtile loaded (can we pass via not loaded tile on the way?)
-    if (!_navMesh || !_navMeshQuery || _sourceUnit->HasUnitState(UNIT_STATE_IGNORE_PATHFINDING) ||
+    const Unit* _sourceUnit = _source->ToUnit();
```
Comment (Shauren) [inline src/server/game/Movement/PathGenerator.cpp]: const placement in many places in this PR

Comment (jackpoz) [reply src/server/game/Movement/PathGenerator.cpp]: would "Unit const* const" be better ?

Comment (Shauren) [reply src/server/game/Movement/PathGenerator.cpp]: just `Unit const*`

**Comment by Shauren** [comment]
Needs `class WorldObject;` forward declaration in header to fix build

I don't see this breaking anything either, PathGenerator mostly uses functions available in WorldObject anyway

**Pull request merged**